cassandra/cluster.py

```diff
@@ -5105,6 +5130,49 @@ def __str__(self):
     __repr__ = __str__
 
 
+class PrepareFuture(ResponseFuture):
+    _final_prepare_result = _NOT_SET
+
+    def __init__(self, session, query, keyspace, custom_payload, timeout, prepare_on_all_hosts):
+        super().__init__(session, PrepareMessage(query=query, keyspace=keyspace), None, timeout)
+        self.query_string = query
+        self._prepare_on_all_hosts = prepare_on_all_hosts
+        self._keyspace = keyspace
+        self._custom_payload = custom_payload
+
+    def _set_final_result(self, response):
+        session = self.session
+        cluster = session.cluster
+        prepared_statement = PreparedStatement.from_message(
+            response.query_id, response.bind_metadata, response.pk_indexes, cluster.metadata, self.query_string,
+            self._keyspace, session._protocol_version, response.column_metadata, response.result_metadata_id,
+            cluster.column_encryption_policy)
+        prepared_statement.custom_payload = response.custom_payload
+        cluster.add_prepared(response.query_id, prepared_statement)
+        self._final_prepare_result = prepared_statement
+
+        if self._prepare_on_all_hosts:
+            # trigger asynchronous preparation of query on other C* nodes,
+            # we are on event loop thread, so do not execute those synchronously
+            session.submit(
+                session.prepare_on_all_nodes,
+                self.query_string, self._current_host, self._keyspace)
```

Comment:
It seems strange to me to have this logic embedded within the future impl like this.  Seems like this should be done when the future is created, something like:

```
# _create_prepare_response_future() in this impl returns a regular ResponseFuture
future = self._create_prepare_response_future(query, keyspace, custom_payload, prepare_on_all_hosts)
if prepare_on_all_hosts:
   # Hand waving about partial application here in order to pass in parameters; the point is we get to a future that
   # calls prepare_on_all_hosts() after we've received a response to our prepare here here
   future = future.then(prepare_on_all_hosts)
future._protocol_handler = self.client_protocol_handler
```

Problem of course is the ResponseFuture doesn't support this then() syntax.  It does have native support for callbacks but that isn't the same; that's just a function that gets invoked when the operation completes.  We don't return a new future that returns the result of the function defined in the then() call like you do in most future APIs.

I'm wondering if we can either (a) add something like that or (b) find a way to wrap this functionality in another future lib in order to simplify an impl like this.